src/common/bitset_set.h

```diff
@@ -0,0 +1,442 @@
+ *
+ * Author: Alex Ainscow
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+/* The standard bitset library does not behave like a "std::set", making it
+ * hard to use as a drop-in replacement. This templated class is intended to
+ * behave like a std::set() with the restriction that it can only store values
+ * less than N. The contents is stored as a bitset for efficiency and there are
+ * some extensions (such as insert_range) that exploit the implementation,
+ * however the main intent is that this is a drop-in replacement for std::set.
+ *
+ * The Key must cast to/from int8_t unambiguously and support the pre-increment
+ * operator.
+ */
+
+#pragma once
+#include <cstdint>
+
+#include "include/buffer.h"
+
+template<size_t NumBitsV, typename KeyT>
+class bitset_set {
+  enum {
+    BITS_PER_UINT64 = 64, WORDS = NumBitsV / 64, MAX = NumBitsV
+  };
+
+  // N must be a multiple of 64 and greater than 0.
+  // There is no real limit to size.
+  static_assert(NumBitsV % BITS_PER_UINT64 == 0);
+  static_assert(NumBitsV > 0);
+
+  // Not possible to create a non-const iterator!
+ public:
+  class const_iterator {
+    const bitset_set *set;
+    KeyT pos;
+
+   public:
+    using difference_type = std::int64_t;
+
+    const_iterator() : set(nullptr), pos(0) {
+    }
+
+    const_iterator(const bitset_set *_set, size_t _pos) : set(_set), pos(_pos) {
+    }
+
+    const_iterator(const bitset_set *_set, KeyT _pos) : set(_set), pos(_pos) {
+    }
+
+    const_iterator(const bitset_set *_set) : set(_set), pos(MAX) {
+      for (size_t i = 0; i < WORDS; ++i) {
+        size_t p = std::countr_zero(set->words[i]);
+        if (p != BITS_PER_UINT64) {
+          pos = (i * BITS_PER_UINT64) + p;
+          break;
+        }
+      }
+    }
+
+    const_iterator &operator++() {
+      uint64_t v;
+      size_t i = (int(pos) + 1) / BITS_PER_UINT64;
+      int bit = (int(pos) + 1) % BITS_PER_UINT64;
+      while (i < WORDS) {
+        if (bit == BITS_PER_UINT64) {
+          v = set->words[i];
+        } else {
+          v = set->words[i] & -1ULL << bit;
+        }
+        bit = std::countr_zero(v);
+        if (bit != BITS_PER_UINT64) {
+          pos = (i * BITS_PER_UINT64) + bit;
+          return *this;
+        }
+        ++i;
+      }
+      pos = MAX;
+      return *this;
+    }
+
+    const_iterator operator++(int) {
+      const_iterator tmp(*this);
+      ++(*this);
+      return tmp;
+    }
+
+    bool operator==(const const_iterator &rhs) const {
+      return int(pos) == int(rhs.pos);
+    }
+
+    bool operator!=(const const_iterator &rhs) const {
+      return !operator==(rhs);
+    }
+
+    const KeyT &operator*() const {
+      return pos;
+    }
+  };
+
+  static_assert(std::input_or_output_iterator<const_iterator>);
+
+ private:
+  uint64_t words[WORDS];
```

Comment:
std::array? you'll gain the whole set of algorithms and utilities, without compromising the speed.